File: src/main/ruby/g0301_0400/s0394_decode_string/solution.rb
# #Medium #Top_100_Liked_Questions #String #Stack #Recursion #Level_1_Day_14_Stack #Udemy_Strings
# #Big_O_Time_O(n)_Space_O(n) #2023_11_27_Time_58_ms_(84.09%)_Space_211.1_MB_(40.91%)

# @param {String} s
# @return {String}
def decode_string(s)
  @i = 0
  decode_helper(s)
end

private

def decode_helper(s)
  count = 0
  result = ""

  while @i < s.length
    c = s[@i]
    @i += 1

    if letter?(c)
      result += c
    elsif digit?(c)
      count = count * 10 + c.to_i
    elsif c == ']'
      break
    elsif c == '['
      # sub problem
      repeat = decode_helper(s)
      result += repeat * count
      count = 0
    end
  end

  result
end

def letter?(c)
  c =~ /[a-zA-Z]/
end

def digit?(c)
  c =~ /\d/
end

File: src/main/go/g0301_0400/s0394_decode_string/solution.go
package s0394_decode_string

// #Medium #Top_100_Liked_Questions #String #Stack #Recursion #LeetCode_75_Stack
// #Level_1_Day_14_Stack #Udemy_Strings #Big_O_Time_O(n)_Space_O(n)
// #2025_05_10_Time_0_ms_(100.00%)_Space_4.00_MB_(85.31%)

import (
	"strings"
)

type Decoder struct {
	s string
	i int
}

func (d *Decoder) decodeString(s string) string {
	d.s = s
	return d.decode()
}

func (d *Decoder) decode() string {
	count := 0
	var sb strings.Builder

	for d.i < len(d.s) {
		c := d.s[d.i]
		d.i++

		if isLetter(c) {
			sb.WriteByte(c)
		} else if isDigit(c) {
			count = count*10 + int(c-'0')
		} else if c == ']' {
			break
		} else if c == '[' {
			// subproblem
			repeat := d.decode()
			for count > 0 {
				sb.WriteString(repeat)
				count--
			}
		}
	}

	return sb.String()
}

func isLetter(c byte) bool {
	return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z')
}

func isDigit(c byte) bool {
	return c >= '0' && c <= '9'
}

func decodeString(s string) string {
	d := Decoder{}
	return d.decodeString(s)
}
